s3/s3_transport/s3_transport_test.cpp

```diff
@@ -0,0 +1,601 @@
+    s3_config.debug_flag = debug_flag;
+    s3_config.multipart_flag = use_multipart_flag;
+    s3_config.shared_memory_timeout_in_seconds = 60;
+
+    s3_transport tp1{s3_config};
+    odstream ds1{tp1, s3_prefix + filename};
+
+    if (!ds1.is_open()) {
+        printf("[%d] Open failed.  Exiting...\n", thread_number);
+        return;
+    }
+
+    ds1.seekp(start);
+
+    // doing multiple writes of 1MiB here just to test that that works
+    /*const uint64_t max_write_size = 512*1024*1024;
+    uint64_t write_offset = 0;
+    while (write_offset < current_buffer_size) {
+        uint64_t write_size = std::min(max_write_size, current_buffer_size - write_offset);
+        ds1.write(current_buffer + write_offset, write_size);
+        write_offset += write_size;
+    }*/
+
+    ds1.write(current_buffer, current_buffer_size);
+
+    printf("WRITE DONE FOR %d\n", thread_number);
+
+    // will be automatic
+    ds1.close();
+    printf("CLOSE DONE FOR %d\n", thread_number);
+
+    /*****************************************/
+
+
+    // s3FileWrite copies its buffer so that iRODS can delete it after
+    // s3FileWrite returns
+    delete[] current_buffer;
+
+    ifs.close();
+
+}
+
+void download_part(int thread_number,
+                   const int thread_count,
+                   const uint64_t file_size,
+                   const bool debug_flag,
+                   const char *hostname,
+                   const char *bucket_name,
+                   const char *access_key,
+                   const char *secret_access_key,
+                   const char *filename)
+{
+    printf("%s(%d, %d, %lu, %d, %s, %s, %s, %s, %s)\n",
+          __FUNCTION__,
+          thread_number,
+          thread_count,
+          file_size,
+          debug_flag,
+          hostname,
+          bucket_name,
+          access_key,
+          secret_access_key,
+          filename);
+
+    printf("%s:%d (%s) [download thread=%u, seq=%u] reading from s3 into file \n",
+            __FILE__, __LINE__, __FUNCTION__, thread_number, thread_number);
+
+    // thread in irods only deal with sequential bytes.  figure out what bytes this
+    // thread deals with
+    size_t start = thread_number * (file_size / thread_count);
+    size_t end = 0;
+    if (thread_number == thread_count - 1) {
+        end = file_size;
+    } else {
+        end = start + file_size / thread_count;
+    }
+
+    // open output stream for test
+    std::ofstream ofs;
+    ofs.open((std::string(filename) + std::string(".downloaded")).c_str(),
+            std::ios::out | std::ios::binary);
+
+    if (!ofs.good()) {
+        fprintf(stderr, "failed to open file %s\n", filename);
+        return;
+    }
+
+    size_t current_buffer_size = end - start;
+    char *current_buffer = static_cast<char*>(malloc(current_buffer_size * sizeof(char)));
+
+    /*****************************************
+     * This part actually goes in S3 plugin. *
+     *****************************************/
```

Comment:
again

Reply:
Program has been replaced by catch2 unit tests and deleted.